### PR TITLE
feat: Implement wait for `ctl` commands

### LINKED ISSUE #91: Implement wait flag for ctl commands
Imperatively interacting with a wasmcloud host, or a lattice of interconnected hosts, it's important to be able to know when certain events take place. For example, when you issue this command to a wasmcloud host:
```
wash ctl start actor wasmcloud.azurecr.io/subscriber:0.2.0
```
You receive an acknowledgment that the actor start event has been _received_. All this means is that your command was successfully delivered to the host and that the host is going to proceed with pulling the actor from our OCI registry and starting it. What you don't get from this acknowledgment is the assurance that the actor actually _was_ started.

I propose a `--wait` flag that can be appended to the `start`, `stop`, `link` and `update` subcommands under `ctl`. These are all commands that return with an acknowledgment but not a confirmation of the command executing. When the `--wait` flag is supplied, the command will instead block until the appropriate event can be observed (e.g. ActorStarted for `start actor`).

In order to monitor the event stream, there are two possibilities:
- When connected to hosts running in a lattice via NATS, events are published as NATS messages that can be monitored. This can be used to monitor events when connected via NATS.
- When connected to a host without NATS (which will be a possibility in the REPL after #92 ) the host can be built with an event sender that can be monitored for the same events that would be sent across NATS. 

Both of these possibilities are important. When using `wash` as a command-line interface, we'll want to monitor events that are flowing over NATS messages, as the control interface operates on a lattice that's connected to via NATS. We'll want to do the same thing when `wash up` is launched with a NATS connection. When the REPL is launched without a NATS connection (after #92 ) we should directly build a host with an event receiver that can be used to monitor these events.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #91.

In this PR I am implementing a way for all `ctl` commands to wait for the outcome of the command, instead of just returning after acknowledgement. This will modify all of these commands to wait by default, with a `--skip-wait` flag that can be passed to mimic the current behavior.

So far I've done this for `ctl start actor`. I'd like to hear feedback on this implementation before I go on to the rest of the commands.

Signed-off-by: Matt Wilkinson <mattwilki17@gmail.com>

### CODE REVIEW

File: src/ctl/mod.rs [hunk 1]
```diff
@@ -328,6 +331,15 @@ pub(crate) struct StartActorCommand {
     /// Timeout to await an auction response, defaults to 2000 milliseconds
     #[clap(long = "auction-timeout-ms", default_value_t = default_timeout_ms())]
     auction_timeout_ms: u64,
+
+    /// By default, the command will wait until the actor has been started.
+    /// If this flag is passed, the command will return immedately after acknowledgement from the host, without waiting for the actor to start.
```
Comment (brooksmtownsend) [inline src/ctl/mod.rs]: nit: typo in `immediately`

File: src/ctl/mod.rs [hunk 2]
```diff
@@ -328,6 +331,15 @@ pub(crate) struct StartActorCommand {
...
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await an actor start, defaults to 10000 milliseconds.
+    #[clap(long = "start-timeout-ms", default_value_t = 10000)]
+    start_timeout_ms: u64,
```
Comment (brooksmtownsend) [inline src/ctl/mod.rs]: I feel like this is a little long, but I think we can hone in on a proper "normal" timeout as we test this out a bit. If something like 3000ms covers the 99% of actor starts that should be plenty, and since we listen for actor start / actor start failed it should still be very quick.

File: src/ctl/mod.rs [hunk 3]
```diff
@@ -715,10 +720,37 @@ pub(crate) async fn start_actor(mut cmd: StartActorCommand) -> Result<CtlOperati
...
+        .map_err(convert_error)?;
+
+    if !ack.accepted {
+        bail!("Operation failed: {}", ack.error);
+    }
+
+    if cmd.skip_wait {
+        return Ok(CommandOutput::from_key_and_text(
+            "result",
+            format!("Starting actor {} on host {}", &cmd.actor_ref, &host),
```
Comment (brooksmtownsend) [inline src/ctl/mod.rs]: To avoid confusion, I prefer the `Start actor request received:` language in the original, that way it's explicit that the host received and acknowledged the request. I like the idea of including the host ID like you did here.

File: src/ctl/mod.rs [hunk 4]
```diff
@@ -715,10 +720,37 @@ pub(crate) async fn start_actor(mut cmd: StartActorCommand) -> Result<CtlOperati
...
+    }
+
+    let start_actor_result_checker =
+        find_start_actor_result_event(host.to_string(), cmd.actor_ref.clone());
+
+    find_event(
+        &receiver,
+        Duration::from_millis(cmd.start_timeout_ms),
+        start_actor_result_checker,
+    )??;
```
Comment (brooksmtownsend) [inline src/ctl/mod.rs]: I feel like it would be more ergonomic to do something like
```
let start_actor_result = wait_for_actor_start_event(receiver, timeout, host, actor_ref)?;
```

And instead wrap the `find_event` logic within the `wait_for_actor_start_event` function. See other comment below for more

Comment (brooksmtownsend) [reply src/ctl/mod.rs]: Of course, feel free to rename the function as you see fit, it's long but descriptive

File: src/ctl/wait.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        let outcome = check_function(event)?;
+
+        if let Some(result) = outcome {
+            return Ok(result);
+        }
+    }
+}
+
+pub(crate) fn find_start_actor_result_event(
```
Comment (brooksmtownsend) [inline src/ctl/wait.rs]: As mentioned above, I'd rather see this function take the receiver and the timeout and call find_event internally. I'm fine with this function passing this closure to `find_event`

File: src/ctl/wait.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                    .ok_or(anyhow!("No image_ref found in data"))?
+                    .as_str()
+                    .ok_or(anyhow!("image_ref is not a string"))?
+                    .to_string();
+
+                if image_ref == actor_ref {
+                    return Ok(Some(Ok(())));
+                }
+            }
+            "com.wasmcloud.latticeactor_start_failed" => {
```
Comment (brooksmtownsend) [inline src/ctl/wait.rs]: This needs to be `com.wasmcloud.lattice.actor_start_failed`

File: src/ctl/wait.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        if let Some(result) = outcome {
+            return Ok(result);
+        }
+    }
+}
+
+pub(crate) fn find_start_actor_result_event(
+    host_id: String,
+    actor_ref: String,
+) -> impl Fn(Event) -> Result<Option<Result<()>>> {
```
Comment (brooksmtownsend) [inline src/ctl/wait.rs]: So here's the scenarios that I see in this function
1. Event received that isn't relevant
2. Error coercing event types or unwrapping its fields
3. Matching event received (success)
4. Matching event received (failure)

Because of this, I think we could simplify this using an enum where the variants are
1. Success
1. Failed
1. NotApplicable

That way we return an `Err` when we get unexpected things when deserializing or unwrapping, and `Ok` with the enum variant that describes that event. Curious what you think about simplifying this Return type a bit

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
This is a great start, and handles the various cases well for starting an actor. I had some code ergonomics suggestions and some minor nits

File: src/ctl/wait.rs (hunk 6 above)
Comment (ghost) [reply src/ctl/wait.rs]: Woops. Will fix

File: src/ctl/wait.rs (hunk 7 above)
Comment (ghost) [reply src/ctl/wait.rs]: So you are suggesting something like `Result<ResponseEnum>`? Where `ResponseEnum` has those three variants. Note ResponseEnum is a working title, I'll come up with a better one

File: src/ctl/mod.rs (hunk 4 above)
Comment (ghost) [reply src/ctl/mod.rs]: Yeah I like that approach. Will work on implementing it

File: src/ctl/mod.rs (hunk 2 above)
Comment (ghost) [reply src/ctl/mod.rs]: Yeah I wasn't sure what the best timeout was, just stuck 10sec in there to be safe to start with. I can drop it to 3000

File: src/ctl/wait.rs (hunk 7 above)
Comment (brooksmtownsend) [reply src/ctl/wait.rs]: That's what I could think of to simplify the cases a little bit, yes

**Comment by ghost** [comment]
@brooksmtownsend I've made all of the changes from your feedback. Could you take a look when you have a bit? Thanks!

**Review by brooksmtownsend — DISMISSED** [review]
Changes look great! I'd give a 👍🏻 to proceed with provider start too

**Comment by ghost** [comment]
I've finished adding the wait parameters to the commands that can support them right now (that have `CloudEvents` being sent over the `wasmbus.evt.*` channel), and rebased to pick up the spinner changes. However the CI is failing on a weird fmt error:

```
Warning: can't set `imports_granularity = Crate`, unstable features are only available in nightly channel.
```

Is this new? Can we just turn that off in the `rustfmt.toml`?

**Comment by ghost** [comment]
@brooksmtownsend this is ready for review

File: src/ctl/mod.rs [hunk 8]
```diff
@@ -399,6 +432,15 @@ pub(crate) struct StopProviderCommand {
...
     #[clap(name = "contract-id")]
     pub(crate) contract_id: String,
+
+    /// By default, the command will wait until the provider has been stopped.
+    /// If this flag is passed, the command will return immediately after acknowledgement from the host, without waiting for the provider to stop.
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await the provider stop, defaults to 3000 milliseconds.
+    #[clap(long = "start-timeout-ms", default_value_t = 3000)]
```
Comment (stevelr) [inline src/ctl/mod.rs]: the command is `stop provider` but this arg is `start-timeout-ms`. I think it would be simpler and less for people to remember if we use `timeout-ms` for all timeout args (unless there's a conflict within the same sub-command and a prefix is needed to disambiguate between two separate timeouts)

File: src/ctl/mod.rs [hunk 9]
```diff
@@ -377,6 +401,15 @@ pub(crate) struct StopActorCommand {
...
     #[clap(long = "count", default_value = "1")]
     pub(crate) count: u16,
+
+    /// By default, the command will wait until the actor has been stopped.
+    /// If this flag is passed, the command will return immediately after acknowledgement from the host, without waiting for the actor to stp[].
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await rhe actor stop, defaults to 3000 milliseconds.
+    #[clap(long = "start-timeout-ms", default_value_t = 3000)]
```
Comment (stevelr) [inline src/ctl/mod.rs]: change to `long = "timeout-ms"`
(should have been `stop-timeout-ms` but I think dropping the start/stop prefix is preferable)

File: src/ctl/mod.rs [hunk 10]
```diff
@@ -359,6 +374,15 @@ pub(crate) struct StartProviderCommand {
...
     #[clap(long = "config-json")]
     config_json: Option<PathBuf>,
+
+    /// By default, the command will wait until the provider has been started.
+    /// If this flag is passed, the command will return immediately after acknowledgement from the host, without waiting for the provider to start.
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await the provider start, defaults to 3000 milliseconds.
+    #[clap(long = "start-timeout-ms", default_value_t = 3000)]
```
Comment (stevelr) [inline src/ctl/mod.rs]: change to long = "timeout-ms"

File: src/ctl/mod.rs [hunk 11]
```diff
@@ -328,6 +334,15 @@ pub(crate) struct StartActorCommand {
...
     #[clap(long = "auction-timeout-ms", default_value_t = default_timeout_ms())]
     auction_timeout_ms: u64,
+
+    /// By default, the command will wait until the actor has been started.
+    /// If this flag is passed, the command will return immediately after acknowledgement from the host, without waiting for the actor to start.
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await an actor start, defaults to 3000 milliseconds.
+    #[clap(long = "start-timeout-ms", default_value_t = 3000)]
```
Comment (stevelr) [inline src/ctl/mod.rs]: change to long = "timeout-ms"

File: src/ctl/mod.rs [hunk 12]
```diff
@@ -377,6 +401,15 @@ pub(crate) struct StopActorCommand {
     /// Number of actors to stop
     #[clap(long = "count", default_value = "1")]
     pub(crate) count: u16,
+
+    /// By default, the command will wait until the actor has been stopped.
+    /// If this flag is passed, the command will return immediately after acknowledgement from the host, without waiting for the actor to stp[].
+    #[clap(long = "skip-wait")]
+    skip_wait: bool,
+
+    /// Timeout to await rhe actor stop, defaults to 3000 milliseconds.
```
Comment (brooksmtownsend) [inline src/ctl/mod.rs]: nit: typo `await rhe actor`

File: src/ctl/wait.rs [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+/// with the `FindEventOutcome` enum containing the success or failure state of the event.
+///
+/// If the timeout is reached or another error occurs, the `Err` variant of the `Result` will be returned.
+///
+/// You can use the generics in `EventCheckOutcome` and `FindEventOutcome` to return any data from the event out of your `check_function`.
+fn find_event<T>(
+    receiver: &Receiver<Event>,
+    timeout: Duration,
+    check_function: impl Fn(Event) -> Result<EventCheckOutcome<T>>,
+) -> Result<FindEventOutcome<T>> {
```
Comment (brooksmtownsend) [inline src/ctl/wait.rs:20]: One thing I'm considering would be useful here, especially for the actor start/stop events, is a `count` parameter where you can wait for a specified number of events. 

For example, if you do a `wash ctl start actor <echo> --count 5`, the ideal scenario would be to wait for 5 `ActorStarted` events. 

I'm personally willing to file that as a feature request though and it doesn't need to be done in this PR. I think I lean towards the feature request only because waiting for multiple events introduces a few questions that I don't have a good answer to, e.g.
1. What if you receive 3/5 successful events before the timeout, is that a mixed success or a fail?
2. What if you receive 99 successful events and 1 failure?
3. What if you receive 1 successful event and 99 failures?

TL;DR I think we feature request awaiting multiple as there are some nuances

**Review by brooksmtownsend — COMMENTED** [review]
Overall this looks awesome and I'm installing it to test now... just left one nit for a typo and a larger comment meant for some discussion

File: src/ctl/mod.rs (hunk 10 above)
Comment (brooksmtownsend) [reply src/ctl/mod.rs]: For starting a provider, this timeout may need to be a good bit larger as it's a larger file to download. I'd probably set the default at 15 seconds, in the general case this will be enough time and in the constrained device case we can always set it higher.

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
image

Works like a charm! Requesting one change regarding the start provider timeout, and agree with Steve's suggestions.

File: src/ctl/mod.rs (hunk 8 above)
Comment (ghost) [reply src/ctl/mod.rs]: The problem is that the `ConnectionOps` struct already has a `timeout_ms` value, which indicates the timeout for receiving the ack from the lattice: https://github.com/wasmCloud/wash/blob/main/src/ctl/mod.rs#L60-L67

I can fix it so it is 'start' and 'stop' in the correct places though

File: src/ctl/wait.rs (hunk 13 above)
Comment (ghost) [reply src/ctl/wait.rs:20]: Yeah that would be a good feature, I agree it should probably be handled in another PR though

File: src/ctl/mod.rs (hunk 8 above)
Comment (stevelr) [reply src/ctl/mod.rs]: Then I would make it so timeout-ms is the timeout for the operation, and ack-timeout-ms (or connect-timeout-ms) is timeout for the connection Ack. 

More than 99% use case would be timeout for the operation, so that should be the simpler form

**Review by brooksmtownsend — DISMISSED** [review]
My feedback was addressed, @mattwilkinsonn you can go ahead and bump this to `0.10.0-alpha.2` so we can release an alpha version once this merges 👍🏻 

@stevelr any other feedback?

**Comment by ghost** [comment]
Last thing - is there a reason we disable the help flag on these ctl commands? Seems rather user unfriendly. I see that there's a few flags with a `-h` short option, could we remove those and enable all of the helps?

Can do this in another PR though

**Comment by brooksmtownsend** [comment]
> Last thing - is there a reason we disable the help flag on these ctl commands? Seems rather user unfriendly. I see that there's a few flags with a -h short option, could we remove those and enable all of the helps?

I wasn't aware we did this... we can keep that separate from this PR but every command should support `-h` and `--help`

File: src/ctl/mod.rs (hunk 8 above)
Comment (ghost) [reply src/ctl/mod.rs]: Fixed, made the action timeout `timeout_ms`, and the ack timeout `ack_timeout_ms`.

**Comment by ghost** [comment]
All feedback is resolved @brooksmtownsend @stevelr. Should be ready to merge

**Comment by ghost** [comment]
Also - the tests didn't run on my last commit? Is there something weird going on there? It only checked the signoff

EDIT: nvm, looks like they ran now

**Comment by stevelr** [comment]
LGTM

**Review by brooksmtownsend — APPROVED** [review]
Awesome work again @mattwilkinsonn !
